scripts/git_changes_utils.py

```diff
@@ -0,0 +1,332 @@
+            'To do that follow these steps:\n'
+            '1. Run the command \'git remote -v\'\n'
+            '2. This command will list the remote references. There will be '
+            'multiple remotes listed for \'upstream\', but we want to make sure'
+            ' that there is only one main \'upstream\' remote. Please use the '
+            'command, \'git remote remove <remote_name>\' on all remotes '
+            'that are not the main oppia repository.\n'
+        )
+        return None
+    return remote_name
+
+
+def git_diff_name_status(
+    left: Optional[str] = None, right: Optional[str] = None,
+    diff_filter: Optional[str] = None
+) -> List[FileDiff]:
+    """Compare two branches/commits with git.
+
+    Parameter:
+        left: str. The name of the lefthand branch.
+        right: str. The name of the righthand branch.
+        diff_filter: str. Arguments given to --diff-filter (ACMRTD...).
+
+    Returns:
+        list. List of FileDiffs (tuple with name/status).
+
+    Raises:
+        ValueError. Raise ValueError if git command fails or if invalid
+            arguments are provided.
+    """
+    git_cmd = ['git', 'diff', '--name-status']
+    if diff_filter == '':
+        raise ValueError(
+            'Error: diff_filter should not be an empty string.')
+    if diff_filter:
+        git_cmd.append('--diff-filter={}'.format(diff_filter))
+    if left == '':
+        raise ValueError('Error: left should not be an empty string.')
+    if right == '':
+        raise ValueError('Error: right should not be an empty string.')
+    if left and right:
+        git_cmd.extend([left, right])
+        # Append -- to avoid conflicts between branch and directory name.
+        # More here: https://stackoverflow.com/questions/26349191
+        git_cmd.append('--')
+    out, err = common.start_subprocess_for_result(git_cmd)
+    if not err:
+        file_list = []
+        for line in out.splitlines():
+            # The lines in the output generally look similar to these:
+            #
+            #   A\tfilename
+            #   M\tfilename
+            #   R100\toldfilename\tnewfilename
+            #
+            # We extract the first char (indicating the status), and the string
+            # after the last tab character.
+            file_list.append(
+                FileDiff(bytes([line[0]]), line[line.rfind(b'\t') + 1:]))
+        return file_list
+    else:
+        raise ValueError(err)
+
+
+def check_file_inside_directory(file_path: str, directory_path: str) -> bool:
+    """Checks if a file is inside a directory.
+
+    Args:
+        file_path: str. The path to the file.
+        directory_path: str. The path to the directory.
+
+    Returns:
+        bool. Whether the file is inside the directory.
+    """
+    file_path = os.path.abspath(file_path)
+    directory_path = os.path.abspath(directory_path)
+
+    return bool(
+        os.path.commonpath([file_path, directory_path]) == directory_path
+    )
```

Comment:
Do you need the outer `bool(...)`? I think an `==` operation between 2 strings will always produce a bool, right?

Reply:
I first tried without bool(...), but mypy complained about how it was returning of type `Any` instead of `bool`, adding this fixed it.

Reply:
Investigated more, even tried separating the os.path.commonpath into another variable (showing both are strings) and the equals should convert the expression into a bool, but mypy still complains.

Reply:
I checked the type hints for `os.path` in [typeshed](https://github.com/python/typeshed/blob/main/stdlib/posixpath.pyi), and I think I know what's going on here. You use `os.path.abspath()` to generate both `file_path` and `directory_path`, and `os.path.abspath()` has a return type of `AnyStr` ([code source](https://github.com/python/typeshed/blob/f54427b660a2f78559cdeb8915a897adec25e832/stdlib/posixpath.pyi#L80)). Note that `AnyStr` refers to a variable that could be either `str` or `bytes` ([mypy docs](https://mypy.readthedocs.io/en/stable/generics.html#type-variables-with-value-restriction)). `commonpath()` doesn't have type hints for inputs of type `AnyStr`, only for `LiteralString`, `str`, or `bytes`. In other words, the type hints for `commonpath()` don't allow for lists that mix `str` and `bytes`, which makes sense. The problem is that since `file_path` and `directory_path` have types `AnyStr`, one could be of type `str` while the other is of type `bytes`. I think mypy is getting confused by this.

Since we know the arguments to the function are strings, we know `file_path` and `directory_path` should also be strings. If you give them type hints explicitly, does that fix the issue? Something like this:

```python
file_path: str = os.path.abspath(file_path)
directory_path: str = os.path.abspath(directory_path)

return os.path.commonpath([file_path, directory_path]) == directory_path
```

Sorry if you already tried this--I couldn't quite tell from your comment what exactly you tried already. If that doesn't work, I'd try the following:

Explicitly declare the commonpath output to be a string:

```python
file_path: str = os.path.abspath(file_path)
directory_path: str = os.path.abspath(directory_path)
common: str = os.path.commonpath([file_path, directory_path])

return common == directory_path
```

Check comparing two plain strings as a sanity check:

```python
a = 'a'
b = 'b'
return a == b
```

Reply:
Thanks that worked, Done.